### ISSUE TITLE
[FEATURE] 204 No Content support for OpenAPI schema

### ISSUE THREAD

**Issue by dmontagu** [tiangolo/fastapi#259]
I would like to be able to specify in the endpoint decorator that the endpoint will have no response (204 No Content). This is documented in the [swagger.io specification](https://swagger.io/docs/specification/describing-responses/) under the heading Empty Response Body.

**Is your feature request related to a problem? Please describe.**
This would allow my `openapi-generator`-generated client to detect that there will not be a response, and will automatically make use of an appropriate response (non-)decoder. 

It's not a terribly hard issue to work around manually by tweaking my generated client, but it would simplify things if I could generate an OpenAPI schema without the content field for the 204-response endpoints.

**Describe the solution you'd like**
So far, my favorite version of an end-user-facing API would involve introducing a new response class (perhaps called `NoContentResponse`) to pass in the endpoint decorator. 

(But I'm not sure how that would interact with `starlette` or if detecting the response_class is even feasible when generating the OpenAPI spec.)

**Comment by tiangolo** [comment]
I see. I'll check it.

**Comment by b1-luettje** [comment]
Any news on this issue? I'm facing the same problem and tried something like

```python
        HTTP_204_NO_CONTENT: dict(
            description="Successful action.",
            content=None,
        )
```
Which only lead to `AttributeError: 'NoneType' object has no attribute 'setdefault'`.

This seemed like an intuitive way to me, but it would be even more intuitive if no content was the default for a 204 with a warning for the user if he declared `content` explicitly.

**Comment by brotskydotcom** [comment]
I just ran into this as well.  I have an idea about how to make `get_openapi_path` do the right thing if it sees no media type on the `response_class` or `reponses` keyword args.  That would be backward-compatible with existing code, and would a decorator such as:
```python
@app.delete('/frotz/{fid}', status_code=204, response_class=Response)
```
or
```python
@app.post('/frotz/', 
          status_code=201,
          responses={302: {'description': 'Such a frotz already exists', content: None})
```
to do the right thing.

**Comment by reynoldsdj** [comment]
I ran into this as well for status code 204s.

**Comment by brotskydotcom** [comment]
@tiangolo any chance you could get time to review my proposed fix?  It’s pretty straightforward. -d.

**Comment by QasimK** [comment]
FastAPI is wonderful, and it handles many common situations when developing APIs, except, seemingly, an endpoint that may or may not return an object.

I have a very simple endpoint to retrieve an object

```py
@app.get("/", response_model=MyModel)
async def root(obj_id: int):
    return await get_object(obj_id)
```

The object does not exist, in which case the function would return `None`.

I would like FastAPI to return HTTP 204 No Content, and for this to be reflected in the documentation.

Right now, I have this workaround:

```py
@app.get("/", response_model=MyModel)
async def root(obj_id: int):
    obj = await get_object(obj_id)
    if obj is None:
        return Response(status_code=HTTP_204_NO_CONTENT)
    return obj
```

However, this is not reflected in the OpenAPI documentation. The documentation does not say that HTTP 204 with an empty HTTP payload is a valid response.

**Comment by brotskydotcom** [comment]
Actually this issue was fixed in #621 and released in [0.43.0](https://github.com/tiangolo/fastapi/releases/tag/0.43.0).  (@tiangolo you might want to close this issue.)

To get your additional responses documented, be sure to include them in your `@app.get` annotation, as described in [this tutorial section](https://fastapi.tiangolo.com/advanced/additional-responses/).  In this case, you would want to list 204 as one of the additional responses.  Note, however, that it is more conventional to return 404 (Not Found) as the response when a user requests a resource that doesn't exist.

**Comment by QasimK** [comment]
Thanks @brotskydotcom, that resolved my issue!

For my use-case, I was simply did this:

```
@app.get("/", response_model=MyModel, responses={404: {"model": None}})
async def root(obj_id: int):
    obj = await get_object(obj_id)
    if obj is None:
        return Response(status_code=HTTP_404_NOT_FOUND)
    return obj
```

I think it would be nice if FastAPI supported this without the boilerplate, as I think it is standard to return 404 when there is no resource (i.e. `return None`). However, perhaps this is just my perspective on things?

**Comment by brotskydotcom** [comment]
It's all about whether you think it's worth documenting 404 as a possible response from any API.  FastAPI gives you the choice, rather than forcing your hand.

**Comment by dmontagu** [comment]
@QasimK We've looked into various ways to automatically detect that a function may raise an HTTPException with a specific status code; the problem is that it is very difficult (impossible in general) to statically determine all of the errors that could be raised from inside a specific function.

You can reduce the boilerplate to some degree by creating a reused dict to use as the responses argument (you could use a function if you need the set of responses to be parametric in some way). But sadly I don't think there is currently a good way to make this work in general.

I would love to see it if someone figures out how to make this work cleanly using `ast` / `typed_ast`  / `mypy` internals or similar, but I'm sure it would be a LOT of work!

**Comment by tiangolo** [comment]
This should be fixed as part of #621, thanks for the report @dmontagu , and thanks for the fix @brotskydotcom !

Also thanks everyone for the discussion here.

As this should now be fixed (from a while ago), I'll close this issue now. But feel free to add more comments or issues.
